src/app/utils/logic-adaptor.ts

```diff
@@ -0,0 +1,95 @@
+import { cloneDeep, omit } from 'lodash'
+
+import * as Logic from '../../shared/util/logic'
+import {
+  BasicField,
+  FieldResponse,
+  ICheckboxFieldSchema,
+  ICheckboxResponse,
+  IFieldSchema,
+  IFormDocument,
+  IPopulatedForm,
+  IPreventSubmitLogicSchema,
+} from '../../types'
+
+const isCheckboxField = (
+  field: IFieldSchema | undefined,
+): field is ICheckboxFieldSchema => {
+  return field?.fieldType === BasicField.Checkbox
+}
+
+const isCheckboxFieldResponse = (
+  field: FieldResponse,
+): field is ICheckboxResponse => {
+  return field.fieldType === BasicField.Checkbox
+}
+
+const convertServerCheckboxValue = (
+  field: ICheckboxResponse,
+  formFields: IPopulatedForm['form_fields'],
+): Logic.LogicFieldResponse => {
+  const completeField = formFields.find(
+    (formField) => String(formField._id) === String(field._id),
+  )
+  if (!isCheckboxField(completeField)) {
+    // eslint-disable-next-line typesafe/no-throw-sync-func
+    throw new Error(`${completeField} is not a Checkbox field`)
+  }
```

Comment:
Understood but using `neverthrow` changes the signatures of these functions which shouldn't be the case for adaptors. If this error is ever thrown, it means that theres inconsistency in our data because the response and field have the same `_id` but have different `fieldTypes` so I thought it would be good to fail early. What do you think? Do you think it would be better to just log the error and return a dummy transformed response? (i.e. empty `options` and false `others`)

Reply:
>so I thought it would be good to fail early

fair enough! let's stick to this then

> using neverthrow changes the signatures of these functions which shouldn't be the case for adaptors

actually isn't the whole point of adapters to wrap underlying functionality in more suitable signatures? so we could add `neverthrow` to the backend adapter? (this is the advantage of separating the adapters btw, looser coupling)